server/src/main/java/org/elasticsearch/index/mapper/CompositeSyntheticFieldLoader.java

```diff
@@ -125,12 +145,30 @@ public interface SyntheticFieldLoaderLayer extends SourceLoader.SyntheticFieldLo
      * Layer that loads malformed values stored in a dedicated field with a conventional name.
      * @see IgnoreMalformedStoredValues
      */
-    public static class MalformedValuesLayer implements SyntheticFieldLoaderLayer {
+    public static class MalformedValuesLayer extends StoredFieldLayer {
+        public MalformedValuesLayer(String fieldName) {
+            super(IgnoreMalformedStoredValues.name(fieldName));
+        }
+
+        @Override
+        protected void writeValue(Object value, XContentBuilder b) throws IOException {
+            if (value instanceof BytesRef r) {
+                XContentDataHelper.decodeAndWrite(b, r);
+            } else {
+                b.value(value);
+            }
+        }
+    }
+
+    /**
+     * Layer that loads field values from a provided stored field.
+     */
+    public abstract static class StoredFieldLayer implements SyntheticFieldLoaderLayer {
```

Comment:
Nit: call it just `Layer`, since it's defined within `CompositeSyntheticFieldLoader`?